akka-sample-kafka-to-sharding-scala/processor/src/main/scala/akka/kafka/KafkaClusterSharding.scala

```diff
@@ -0,0 +1,145 @@
+   * cluster for the number of partitions of a user provided [[topic]]. Use the [[settings]] parameter to configure
+   * the Kafka Consumer connection required to retrieve the number of partitions. Use the [[entityIdExtractor]] to pick
+   * a field from the Entity to use as the entity id for the hashing strategy.
+   *
+   * _Important_: All topics used in a Consumer [[Subscription]] must contain the same number of partitions to ensure
+   * that entities are routed to the same Entity type.
+   */
+  def messageExtractorNoEnvelope[M](system: ActorSystem,
+                                    topic: String,
+                                    timeout: FiniteDuration,
+                                    entityIdExtractor: M => String,
+                                    settings: ConsumerSettings[_,_]): Future[KafkaShardingNoEnvelopeExtractor[M]] =
+    getPartitionCount(system, topic, timeout, settings)
+      .map(kafkaPartitions => new KafkaShardingNoEnvelopeExtractor[M](kafkaPartitions, entityIdExtractor))(system.dispatcher)
+
+  private def getPartitionCount[M](system: ActorSystem, topic: String, timeout: FiniteDuration, settings: ConsumerSettings[_, _]): Future[Int] = {
+    implicit val ec: ExecutionContextExecutor = system.dispatcher
+    val actorNum = metadataActorCounter.getAndIncrement()
+    val consumerActor = system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(KafkaConsumerActor.props(settings), s"metadata-consumer-actor-$actorNum")
+    val metadataClient = MetadataClient.create(consumerActor, timeout)
+    val numPartitions = metadataClient.getPartitionsFor(topic).map(_.length)
+    numPartitions.map { count =>
+      system.log.info("Retrieved {} partitions for topic '{}'", count, topic)
+      count
+    }
+  }
+
+  sealed trait KafkaClusterSharding {
+    def kafkaPartitions: Int
+    def shardId(entityId: String): String = {
+      // simplified version of Kafka's `DefaultPartitioner` implementation
+      val partition = org.apache.kafka.common.utils.Utils.toPositive(Utils.murmur2(entityId.getBytes())) % kafkaPartitions
+      partition.toString
+    }
+  }
+
+  final class KafkaShardingMessageExtractor[M](val kafkaPartitions: Int)
+    extends ShardingMessageExtractor[ShardingEnvelope[M], M] with KafkaClusterSharding {
+    override def entityId(envelope: ShardingEnvelope[M]): String = envelope.entityId
+    override def unwrapMessage(envelope: ShardingEnvelope[M]): M = envelope.message
+  }
+
+  final class KafkaShardingNoEnvelopeExtractor[M](val kafkaPartitions: Int, entityIdExtractor: M => String)
+    extends ShardingMessageExtractor[M, M] with KafkaClusterSharding {
+    override def entityId(message: M): String = entityIdExtractor(message)
+    override def unwrapMessage(message: M): M = message
+  }
+
+  // TODO:
+  // - will require `akka-actors-typed` as another provided dep, or should we just return a classic actor?
```

Comment:
I'm divided. On one hand I'd prefer to start pushing users into typed, but at the same time we should provide a typed version of `Subscriptions.[...].withRebalanceListener(rebalanceListener.toClassic)` otherwise it's a bit inconsistent.

Reply:
External sharding strategies require Akka Typed APIs, but as you say, Alpakka Kafka APIs (except this one) support it yet.  Since setting the Alpakka Kafka rebalance listener requires a classic `ActorRef` this API should return it as such.

Reply:
Since the reference on `actor-typed` is required anyway for Cluster Sharding, I kept the internal implementation as a typed actor, but instead of returning an `Actor[_]` I spawn it and return a classic `ActorRef` that can easily be passed to the Alpakka Kafka `ConsumerSettings`.